include/psa/crypto.h

```diff
@@ -3759,6 +3759,418 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *    #PSA_PAKE_OPERATION_INIT.
+ * -# Call psa_pake_setup() to specify the algorithm, the key, cipher suite,
+ *    identities and additional session information.
+ *
+ * A typical sequence of calls to perform a password-authenticated key
+ * exchange:
+ * -# Call psa_pake_get_key_share() to get the key share that needs to be sent
+ *    to the peer.
+ * -# Call psa_pake_set_key_share() to provide the key share that was received
+ *    from the peer.
+ * -# Call psa_pake_get_implicit_key() for accessing the shared secret.
+ *
+ * The exact sequence of calls to perform a password-authenticated key exchange
+ * depends on the protocol in use:
+ * -# Some protocols exchange more data than just a single key share. When using
+ *    such a protocol, call psa_pake_output() and psa_pake_input() one or more
+ *    times to exchange any further data that is needed to derive the shared
+ *    secret.
+ *
+ * Refer to the documentation of individual PAKE algorithm types (`PSA_ALG_XXX`
+ * values of type ::psa_algorithm_t such that #PSA_ALG_IS_PAKE(\c alg) is true)
+ * for more information.
+ *
+ * If an error occurs at any step after a call to psa_pake_setup(),
+ * the operation will need to be reset by a call to psa_pake_abort(). The
+ * application may call psa_pake_abort() at any time after the operation
+ * has been initialized.
+ *
+ * After a successful call to psa_pake_setup(), the application must
+ * eventually terminate the operation. The following events terminate an
+ * operation:
+ * - A call to psa_pake_abort().
+ * - A successful call to psa_pake_get_implicit_key().
+ *
+ * \param[in,out] operation     The operation object to set up. It must have
+ *                              been initialized as per the documentation for
+ *                              #psa_pake_operation_t and not yet in use.
+ * \param password              Identifier of the key holding the password or a
+ *                              value derived from the password (eg. by a
+ *                              memory-hard function).  It must remain valid
+ *                              until the operation terminates. It must be of
+ *                              type PSA_KEY_TYPE_PASSWORD or
+ *                              #PSA_KEY_TYPE_DERIVE. It has to allow the usage
+ *                              #PSA_KEY_USAGE_DERIVE.
+ * \param alg                   The PAKE protocol to use
+ *                              (\c PSA_ALG_XXX value such that
+ *                              #PSA_ALG_IS_PAKE(\p alg) is true).
+ * \param cipher_suite          The cipher suite to use with the PAKE algorithm.
+ * \param side                  A value of type ::psa_pake_side_t signaling the
+ *                              side of the protocol that is being set up. For
+ *                              more information see the documentation of \c
+ *                              PSA_PAKE_SIDE_XXX constants.
+ * \param[in] user_id           The user ID to authenticate with.
```

Comment:
The length of user ids and the session data can vary across applications independently from the PAKE algorithm used. For most algorithm these are included in the handshake transcript in some way. I wanted to make this a single call, because this way even malloc-free implementations can include rolling hashes in the context instead of the raw ids and session data, which makes the context smaller and eliminates possible limitations on the size of input.

In theory we could achieve the same with individual function calls, but then these would need to be called in a specific order, which might not be the same for different PAKEs.

In summary I am inclined to keeping this a single function call. On the other hand you are making a very good point about providing a way to input not just keys, but raw passwords and key derivation objects.

Adding these would mean having something like this:
```
psa_pake_setup_with_key(&operation, key_id, ...);
psa_pake_setup_with_password(&operation, password, password_len, ...);
psa_pake_setup_with_mhf(&pake_operation, &kdf_operation, password_len, ...);
```

What do you think?